src/Compilers/CSharp/Portable/Lowering/Instrumentation/ModuleCancellationInstrumenter.cs

```diff
@@ -0,0 +1,198 @@
+
+    public override void InterceptCallAndAdjustArguments(
+        ref MethodSymbol method,
+        ref BoundExpression? receiver,
+        ref ImmutableArray<BoundExpression> arguments,
+        ref ImmutableArray<RefKind> argumentRefKindsOpt,
+        bool invokedAsExtensionMethod,
+        SimpleNameSyntax? nameSyntax)
+    {
+        Previous.InterceptCallAndAdjustArguments(ref method, ref receiver, ref arguments, ref argumentRefKindsOpt, invokedAsExtensionMethod, nameSyntax);
+
+        if (arguments is [.., { Type: { } lastArgumentType } lastArgument] &&
+            (argumentRefKindsOpt.IsDefault || argumentRefKindsOpt is [.., RefKind.None]) &&
+            lastArgumentType.Equals(_throwMethod.ContainingType, TypeCompareKind.ConsiderEverything))
+        {
+            // The last argument is a CancellationToken. Replace it with the module-level token.
+            // Keep the previous expression so that side-effects are preserved.
+            arguments = [.. arguments[0..^1], _factory.Sequence([lastArgument], _factory.ModuleCancellationToken())];
+        }
+        else if (FindOverloadWithCancellationToken(method) is { } cancellableOverload)
+        {
+            // The method being invoked does not have a CancellationToken as the last parameter, but there is an overload that does.
+            // Invoke the other overload instead and pass in module-level token. 
+            method = cancellableOverload;
+            arguments = [.. arguments, _factory.ModuleCancellationToken()];
+            argumentRefKindsOpt = argumentRefKindsOpt.IsDefault ? default : [.. argumentRefKindsOpt, RefKind.None];
+        }
+    }
+
+    /// <summary>
+    /// Find an overload whose last parameter is <see cref="CancellationToken"/> and
+    /// the parameter types and ref kinds of all other parameters match those of <paramref name="method"/>.
+    /// </summary>
+    private MethodSymbol? FindOverloadWithCancellationToken(MethodSymbol method)
+    {
+        // It's unlikely that real-world APIs have overloads that differ in dynamic,
+        // but if they do avoid selecting them since their intended use might differ.
+        //
+        // Similarly, we also only consider overloads with the same visibility.
+        // We could potentially allow the visibility to differ,
+        // but we'd need to check if the overload is accessible at the call site.
+
+        const TypeCompareKind TypeComparisonKind = TypeCompareKind.CLRSignatureCompareOptions & ~TypeCompareKind.IgnoreDynamic;
+
+        foreach (var member in method.ContainingType.GetMembers(method.Name))
+        {
+            if (member.IsStatic == method.IsStatic &&
+                member.MetadataVisibility == method.MetadataVisibility &&
+                member is MethodSymbol { Parameters: [.., { RefKind: RefKind.None, Type: { } lastParamType }] parametersWithCancellationToken } overload &&
+                overload.Arity == method.Arity &&
```

Comment:
>overload.Arity == method.Arity [](http://example.com/codeflow?start=16&length=30)

We probably should require a ```MethodKind``` to match as well. For example, we shouldn't substitute an accessor with non-accessor, etc. That is if we want to substitute accessors at all. I think we should be explicit about what method kinds we are going to substitute.